analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs

```diff
@@ -25,212 +25,98 @@
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSymbolAction(
-                c =>
+            context.RegisterSymbolAction(c =>
                 {
                     var innerClassSymbol = (INamedTypeSymbol)c.Symbol;
                     var containerClassSymbol = innerClassSymbol.ContainingType;
-                    if (!innerClassSymbol.IsClass() ||
-                        !containerClassSymbol.IsClass())
+                    if (!innerClassSymbol.IsClassOrStruct() || !containerClassSymbol.IsClassOrStruct())
                     {
                         return;
                     }
 
-                    var members = innerClassSymbol.GetMembers().Where(member => !member.IsImplicitlyDeclared);
-                    foreach (var member in members)
+                    foreach (var member in innerClassSymbol.GetMembers().Where(x => !x.IsImplicitlyDeclared))
                     {
-                        if (member is IPropertySymbol property)
+                        switch (member)
                         {
-                            CheckProperty(c, containerClassSymbol, property);
-                            continue;
-                        }
-
-                        if (member is IFieldSymbol field)
-                        {
-                            CheckField(c, containerClassSymbol, field);
-                            continue;
-                        }
-
-                        if (member is IEventSymbol @event)
-                        {
-                            CheckEvent(c, containerClassSymbol, @event);
-                            continue;
-                        }
-
-                        if (member is IMethodSymbol method)
-                        {
-                            CheckMethod(c, containerClassSymbol, method);
-                            continue;
-                        }
-
-                        if (member is INamedTypeSymbol namedType)
-                        {
-                            CheckNamedType(c, containerClassSymbol, namedType);
+                            case IPropertySymbol property:
+                                CheckFieldPropertyEventOrMethod(property, containerClassSymbol, c, "property");
+                                break;
+                            case IFieldSymbol field:
+                                CheckFieldPropertyEventOrMethod(field, containerClassSymbol, c, "field");
+                                break;
+                            case IEventSymbol @event:
+                                CheckFieldPropertyEventOrMethod(@event, containerClassSymbol, c, "event");
+                                break;
+                            case IMethodSymbol { MethodKind: MethodKind.DeclareMethod or MethodKind.Ordinary } method:
```

Comment:
There are a lot of IMethodSymbols, that can not collide, because they are unnamed (and they don't have a location in code).